src/Http/RouteCollection.php

```diff
@@ -63,19 +68,37 @@ public function delete($path, $name, $handler)
 
     public function addRoute($method, $path, $name, $handler)
     {
-        $routeDatas = $this->routeParser->parse($path);
+        $this->routes[$method][$name] = compact('path', 'handler');
 
-        foreach ($routeDatas as $routeData) {
-            $this->dataGenerator->addRoute($method, $routeData, ['name' => $name, 'handler' => $handler]);
-        }
+        return $this;
+    }
 
-        $this->reverse[$name] = $routeDatas;
+    protected function applyRoutes(): void
+    {
+        foreach ($this->routes as $method => $routes) {
+            foreach ($routes as $name => $route) {
+                $routeDatas = $this->routeParser->parse($route['path']);
 
-        return $this;
+                foreach ($routeDatas as $routeData) {
+                    $this->dataGenerator->addRoute($method, $routeData, ['name' => $name, 'handler' => $route['handler']]);
+                }
+
+                $this->reverse[$name] = $routeDatas;
+            }
+        }
+    }
+
+    public function getRoutes(): array
+    {
+        return $this->routes;
     }
 
     public function getRouteData()
     {
+        if (empty($this->reverse)) {
```

Comment:
so, we would essentially apply the routes on the boot phase by manually resolving the route collection, instead of on first demand. That would technically work, I'm just unsure if it's a good idea to take early resolution into account and explicitly throw an exception just to discourage extension devs.

Reply:
That's a good point... Plus, now that:

- Classes registered via extenders will be resolved when needed instead of at subscribe time
- Subscribers will be resolved after boot

early resolution should be much, much less frequent. Yay extenders!!!